cli/napi/node_api.rs

```diff
@@ -0,0 +1,709 @@
+
+#![deny(unsafe_op_in_unsafe_fn)]
+
+use super::util::get_array_buffer_ptr;
+use super::util::make_external_backing_store;
+use super::util::napi_clear_last_error;
+use super::util::napi_set_last_error;
+use super::util::SendPtr;
+use crate::check_arg;
+use crate::check_env;
+use deno_runtime::deno_napi::*;
+use napi_sym::napi_sym;
+use std::ptr::NonNull;
+use std::sync::atomic::AtomicU8;
+use std::sync::atomic::Ordering;
+
+#[napi_sym]
+fn napi_module_register(module: *const NapiModule) -> napi_status {
+  MODULE_TO_REGISTER.with(|cell| {
+    let mut slot = cell.borrow_mut();
+    let prev = slot.replace(module);
+    assert!(prev.is_none());
+  });
+  napi_ok
+}
+
+#[napi_sym]
+fn napi_add_env_cleanup_hook(
+  env: *mut Env,
+  fun: Option<napi_cleanup_hook>,
+  arg: *mut c_void,
+) -> napi_status {
+  let env = check_env!(env);
+  check_arg!(env, fun);
+
+  let fun = fun.unwrap();
+
+  let mut env_cleanup_hooks = env.cleanup_hooks.borrow_mut();
+  if env_cleanup_hooks
+    .iter()
+    .any(|pair| pair.0 == fun && pair.1 == arg)
+  {
+    panic!("Cleanup hook with this data already registered");
+  }
+  env_cleanup_hooks.push((fun, arg));
+
+  napi_ok
+}
+
+#[napi_sym]
+fn napi_remove_env_cleanup_hook(
+  env: *mut Env,
+  fun: Option<napi_cleanup_hook>,
+  arg: *mut c_void,
+) -> napi_status {
+  let env = check_env!(env);
+  check_arg!(env, fun);
+
+  let fun = fun.unwrap();
+
+  let mut env_cleanup_hooks = env.cleanup_hooks.borrow_mut();
+  // Hooks are supposed to be removed in LIFO order
+  let maybe_index = env_cleanup_hooks
+    .iter()
+    .rposition(|&pair| pair.0 == fun && pair.1 == arg);
+
+  if let Some(index) = maybe_index {
+    env_cleanup_hooks.remove(index);
+  } else {
+    panic!("Cleanup hook with this data not found");
+  }
+
+  napi_ok
+}
+
+type AsyncCleanupHandle = (*mut Env, napi_async_cleanup_hook, *mut c_void);
+
+unsafe extern "C" fn async_cleanup_handler(arg: *mut c_void) {
+  unsafe {
+    let (env, hook, arg) = *Box::<AsyncCleanupHandle>::from_raw(arg as _);
+    hook(env as _, arg);
+  }
+}
+
+#[napi_sym]
+fn napi_add_async_cleanup_hook(
+  env: *mut Env,
+  hook: Option<napi_async_cleanup_hook>,
+  arg: *mut c_void,
+  remove_handle: *mut napi_async_cleanup_hook_handle,
+) -> napi_status {
+  let env = check_env!(env);
+  check_arg!(env, hook);
+
+  let hook = hook.unwrap();
+
+  let handle =
+    Box::into_raw(Box::<AsyncCleanupHandle>::new((env, hook, arg))) as _;
+
+  unsafe {
+    napi_add_env_cleanup_hook(env, Some(async_cleanup_handler), handle);
+  }
+
+  if !remove_handle.is_null() {
+    unsafe {
+      *remove_handle = handle;
+    }
+  }
+
+  napi_clear_last_error(env)
+}
+
+#[napi_sym]
+fn napi_remove_async_cleanup_hook(
+  remove_handle: napi_async_cleanup_hook_handle,
+) -> napi_status {
+  if remove_handle.is_null() {
+    return napi_invalid_arg;
+  }
+
+  let handle = unsafe { &*(remove_handle as *mut AsyncCleanupHandle) };
```

Comment:
This should probably consume `AsyncCleanupHandle` (use `Box::from_raw`)

Reply:
there's some annoyance here regarding free if the cleanup hook already ran and i'm going to rewrite cleanup hooks in a followup anyway, so i think it will be ok to address then.